### PR TITLE
#909 h5py cea

### PR DESCRIPTION
**Pull request by JIMENOFONSECA**

### CODE REVIEW

**Comment by JIMENOFONSECA** [comment]
@Khayatian let me know when you have time to try it.
you will need to have this in the config file:

[demand]
year-to-calculate = 2015
buildings = 
resolution = hourly
format-output = hdf5
use-dynamic-infiltration-calculation = false

the result are the buildigns saved in hdf5 directly

File: cea/demand/demand_main.py [hunk 1]
```diff
@@ -69,60 +71,72 @@ def demand_calculation(locator, weather_path, gv, use_dynamic_infiltration_calcu
...
+    weather_data = epwreader.epw_reader(weather_path)[
+        ['drybulb_C', 'wetbulb_C', 'relhum_percent', 'windspd_ms', 'skytemp_C']]
 
     building_properties, schedules_dict, date = properties_and_schedule(gv, locator)
 
     # in case gv passes a list of specific buildings to simulate.
-    if gv.simulate_building_list:
-        list_building_names = gv.simulate_building_list
-    else:
+    if buildings == []:
```
Comment (daren-thomas) [inline cea/demand/demand_main.py]: a more pythonic way of doing this test is `if not buildings` - this has to do with an empty list, as well as `None`, `0`, `""` and `False` being all "falsy", as far as python is concerned.

File: cea/default.config [hunk 2]
```diff
@@ -142,6 +147,32 @@ destination.help = destination folder to extract the reference case to. Since th
 # has-cooling-season = true
 # has-cooling-season.type = BooleanParameter
 
+buildings =
+buildings.type = ListParameter
+buildings.help = a list of buildings to run the simulation for - leave blank to simulate all buildings
+
+loads =
+loads.type = ListParameter
+loads.help = a list of loads to run the simulation for - leave blank to simulate all loads in demand_writers
```
Comment (daren-thomas) [inline cea/default.config]: is the simulation different if not all loads are selected? (I haven't reviewed the other files, so I don't know yet, but I expect this is just a list of loads to write to the output file?)

I think `MultiChoiceParameter` would be a better choice here, as the interface in ArcGIS would then allow the user to choose them from a list - the ArcGIS interface is dynamically constructed based on the contents of the default.config file...

File: cea/demand/demand_main.py [hunk 3]
```diff
@@ -69,60 +71,72 @@ def demand_calculation(locator, weather_path, gv, use_dynamic_infiltration_calcu
...
         list_building_names = building_properties.list_building_names()
+        print('Running demand calculation for all buildings in the zone')
+    else:
+        list_building_names = gv.simulate_building_list
```
Comment (daren-thomas) [inline cea/demand/demand_main.py]: didn't you just add a `config.demand.buildings` variable to use instead of `gv.simulate_building_list`? (or did I misunderstand?)

File: cea/demand/demand_main.py [hunk 4]
```diff
@@ -69,60 +71,72 @@ def demand_calculation(locator, weather_path, gv, use_dynamic_infiltration_calcu
...
+        print('Running demand calculation for the next buildings=%s' % list_building_names)
 
     # demand
     if multiprocessing and mp.cpu_count() > 1:
-        thermal_loads_all_buildings_multiprocessing(building_properties, date, gv, locator, list_building_names,
-                                                    schedules_dict, weather_data, use_dynamic_infiltration_calculation)
+        print("Using %i CPU's" % mp.cpu_count())
+        calc_demand_multiprocessing(building_properties, date, gv, locator, list_building_names,
```
Comment (daren-thomas) [inline cea/demand/demand_main.py]: I like these new names way better than the old ones!!! :-)

File: cea/demand/demand_main.py [hunk 5]
```diff
@@ -69,60 +71,72 @@ def demand_calculation(locator, weather_path, gv, use_dynamic_infiltration_calcu
...
+                                    schedules_dict, weather_data, use_dynamic_infiltration_calculation)
+    else:
+        calc_demand_singleprocessing(building_properties, date, gv, locator, list_building_names, schedules_dict,
+                                     weather_data, use_dynamic_infiltration_calculation)
+
+    # write yearly totals
+    config = cea.config.Configuration()
```
Comment (daren-thomas) [inline cea/demand/demand_main.py]: this is probably a mistake - the `config` object should be passed in through the `main` function - that allows interfaces (cli, grasshopper, arcgis, tests) to override values in the config object.

File: cea/demand/demand_main.py [hunk 6]
```diff
@@ -69,60 +71,72 @@ def demand_calculation(locator, weather_path, gv, use_dynamic_infiltration_calcu
...
-        totals, time_series = gv.demand_writer.write_totals_csv(building_properties, locator)
-        return totals, time_series
+    return totals, time_series
 
     gv.log('done - time elapsed: %(time_elapsed).2f seconds', time_elapsed=time.clock() - t0)
 
 
 def properties_and_schedule(gv, locator):
-    # this script is called from the Neural network please do not mes up with it!.
+    # this script is called from the Neural network please do not mess it up with it!.
```
Comment (daren-thomas) [inline cea/demand/demand_main.py]: should probably be: "please do not mess with it!"

File: cea/demand/demand_writers.py [hunk 7]
```diff
@@ -25,121 +21,171 @@ class DemandWriter(object):
     - implement the `write_to_csv` method
     """
 
-    def __init__(self, gv):
-        self.gv = gv
-        self.vars_to_print = gv.demand_building_csv_columns
+    def __init__(self, loads, massflows, temperatures):
+        if loads == []:
```
Comment (daren-thomas) [inline cea/demand/demand_writers.py]: use `if not loads` to be more pythonic?

File: cea/demand/demand_writers.py [hunk 8]
```diff
@@ -25,121 +21,171 @@ class DemandWriter(object):
...
+            self.LOAD_VARS = ['QEf', 'QHf', 'QCf', 'Ef', 'Qhsf', 'Qhs', 'Qhsf_lat', 'Qwwf', 'Qww', 'Qcsf',
```
Comment (daren-thomas) [inline cea/demand/demand_writers.py]: using uppercase implies that these are constants, but they're not. we could refactor/rename them to be private variables (e.g.  `self._load_vars`, note the leading underscore. or just `self.load_vars`)

File: cea/demand/metamodel/nn_generator/input_matrix.py [hunk 9]
```diff
@@ -10,14 +10,15 @@
...
+import cea.config
 from cea.demand.demand_main import properties_and_schedule
 from cea.demand.thermal_loads import initialize_inputs
 from cea.utilities.dbfreader import dbf_to_dataframe
 from cea.technologies import controllers
 from cea.utilities import epwreader
 from cea.demand.sensible_loads import calc_I_sol
 import numpy as np
-from cea.demand.metamodel.nn_generator.nn_settings import nn_delay, climatic_variables
+from cea.config.neural_network import nn_delay, climatic_variables
```
Comment (daren-thomas) [inline cea/demand/metamodel/nn_generator/input_matrix.py]: did you create a `cea/config` directory? i'm not sure this line can even work...

File: cea/demand/metamodel/nn_generator/input_matrix.py [hunk 10]
```diff
@@ -359,15 +360,14 @@ def get_array_HVAC_variables(building):
     return array_hvac
 
 
-def run_as_script():
+def main(config):
     gv = cea.globalvar.GlobalVariables()
-    scenario_path = gv.scenario_reference
-    locator = cea.inputlocator.InputLocator(scenario=scenario_path)
+    locator = cea.inputlocator.InputLocator(scenario_path=config.scenario)
     building_name = 'B155066'
```
Comment (daren-thomas) [inline cea/demand/metamodel/nn_generator/input_matrix.py]: ideally, this would be moved to the configuration file, as the `main` function get's called for the script.

File: cea/demand/metamodel/nn_generator/input_matrix.py [hunk 11]
```diff
@@ -10,14 +10,15 @@
...
 from cea.demand.thermal_loads import initialize_inputs
 from cea.utilities.dbfreader import dbf_to_dataframe
 from cea.technologies import controllers
 from cea.utilities import epwreader
 from cea.demand.sensible_loads import calc_I_sol
 import numpy as np
-from cea.demand.metamodel.nn_generator.nn_settings import nn_delay, climatic_variables
+from cea.config.neural_network import nn_delay, climatic_variables
 
 __author__ = "Jimeno A. Fonseca","Fazel Khayatian"
```
Comment (daren-thomas) [inline cea/demand/metamodel/nn_generator/input_matrix.py]: this should be a list. wait-a-minute... does this even compile?

File: cea/demand/metamodel/nn_generator/nn_test_sampler.py [hunk 12]
```diff
@@ -67,17 +69,15 @@ def sampling_single(locator, random_variables, target_parameters, list_building_
     return urban_input_matrix, urban_taget_matrix
 
 
-def run_as_script():
+def main(config):
     gv = cea.globalvar.GlobalVariables()
-    scenario_path = gv.scenario_reference
-    locator = cea.inputlocator.InputLocator(scenario_path=scenario_path)
-    weather_path = locator.get_default_weather()
+    locator = cea.inputlocator.InputLocator(scenario_path=config.scenario)
```
Comment (daren-thomas) [inline cea/demand/metamodel/nn_generator/nn_test_sampler.py]: I think the parameter to `InputLocator` is now called `scenario`

File: cea/default.config (hunk 2 above)
Comment (JIMENOFONSECA) [reply cea/default.config]: good point. I just did the cahnges

File: cea/demand/demand_main.py (hunk 5 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_main.py]: great, I am passing it around

File: cea/demand/demand_main.py (hunk 6 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_main.py]: corrected

File: cea/demand/demand_main.py (hunk 4 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_main.py]: :-) me too!.

File: cea/demand/demand_main.py (hunk 3 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_main.py]: my bad. I corrected it

File: cea/demand/demand_main.py (hunk 1 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_main.py]: chnaged!

File: cea/demand/demand_writers.py (hunk 7 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_writers.py]: fixed

File: cea/demand/demand_writers.py (hunk 8 above)
Comment (JIMENOFONSECA) [reply cea/demand/demand_writers.py]: fixed

File: cea/demand/metamodel/nn_generator/input_matrix.py (hunk 11 above)
Comment (JIMENOFONSECA) [reply cea/demand/metamodel/nn_generator/input_matrix.py]: refactored

File: cea/demand/metamodel/nn_generator/input_matrix.py (hunk 10 above)
Comment (JIMENOFONSECA) [reply cea/demand/metamodel/nn_generator/input_matrix.py]: This is to be seen after we refactor the neural network in a new issue

File: cea/demand/metamodel/nn_generator/nn_test_sampler.py (hunk 12 above)
Comment (JIMENOFONSECA) [reply cea/demand/metamodel/nn_generator/nn_test_sampler.py]: fixed

**Comment by JIMENOFONSECA** [comment]
@daren-thomas I checked all the points as requested. thanks a lot for the thorough review. I learnt a lot during the process.

**Comment by JIMENOFONSECA** [comment]
@daren-thomas what are required statuses?
